<a href="https://colab.research.google.com/github/nazaninmehregan/AI4Goodlab2025/blob/main/CSV_preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data preprocessing: more art than science?

## Contents of this notebook:
<ol>
<li>Load and examine your data</li>
<li>Merging two dataframes</li>
<li>Removing features that you do not need</li>
<li>Making your data machine-readable</li>
<li>Handling not available (NA) and inf data</li>
<li>Removing columns with a standard deviation of 0</li>
<li>Feature scaling</li>
<li>Data visualization</li>
<li>Loading local files into Google Colab</li>
<li>Mini assignment: data visualization</li>
</ol>

## 1. Load and examine your data

In [ ]:
!git clone https://github.com/NadiaBlostein/Project-Prep-Series-02-Data-Preparation
%cd Project-Prep-Series-02-Data-Preparation

Cloning into 'Project-Prep-Series-02-Data-Preparation'...
remote: Enumerating objects: 122, done.
remote: Counting objects: 100% (51/51), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 122 (delta 26), reused 35 (delta 12), pack-reused 71 (from 1)
Receiving objects: 100% (122/122), 122.92 MiB | 42.35 MiB/s, done.
Resolving deltas: 100% (49/49), done.
/content/Project-Prep-Series-02-Data-Preparation


### 1.1 Preliminary examination of your data

In [ ]:
import pandas as pd
import numpy as np
import scipy.stats as stats

df = pd.read_csv("/content/Project-Prep-Series-02-Data-Preparation/data_csv/unrestricted_HCP_behavioral.csv")

Quick quality check: is the dataframe shape what you expect it to be? You should know how many subjects (ie rows) and features (ie columns) you are expecting your dataframe (`df`) to have prior to any downstream analysis.

In [ ]:
print(f"df type: {type(df)}")
print(f"df shape: {df.shape}")

df type: <class 'pandas.core.frame.DataFrame'>
df shape: (1206, 499)


In [ ]:
df.head(4)
# The columns are mostly about brain health, cognition, personality, sleep, emotion, and behavior

Subject Gender    Age  MMSE_Score  PSQI_Score  PSQI_Comp1  PSQI_Comp2  \
0   100004      M  22-25          29           8           1           2   
1   100206      M  26-30          30           6           1           1   
2   100307      F  26-30          29           4           1           0   
3   100408      M  31-35          30           4           1           1   

   PSQI_Comp3  PSQI_Comp4  PSQI_Comp5  ...  Noise_Comp  Odor_Unadj  \
0           2           1           1  ...         5.2      101.12   
1           1           1           1  ...         6.0      108.79   
2           1           0           2  ...         3.6      101.12   
3           0           0           1  ...         2.0      108.79   

  Odor_AgeAdj  PainIntens_RawScore PainInterf_Tscore  Taste_Unadj  \
0       86.45                  2.0              45.9       107.17   
1       97.19                  1.0              49.7        72.63   
2       86.45                  0.0              38.6        71.69   
3       98.04                  2.0              52.6       114.01   

   Taste_AgeAdj  Mars_Log_Score  Mars_Errs  Mars_Final  
0        105.31            1.80        0.0        1.80  
1         72.03            1.84        0.0        1.84  
2         71.76            1.76        0.0        1.76  
3        113.59            1.76        2.0        1.68  

[4 rows x 499 columns]

In [ ]:
df.tail(4)

Subject Gender    Age  MMSE_Score  PSQI_Score  PSQI_Comp1  PSQI_Comp2  \
1202   993675      F  26-30          30           4           1           1   
1203   994273      M  26-30          28           8           1           1   
1204   995174      M  22-25          26           7           1           2   
1205   996782      F  26-30          30           4           1           1   

      PSQI_Comp3  PSQI_Comp4  PSQI_Comp5  ...  Noise_Comp  Odor_Unadj  \
1202           1           0           1  ...         0.4      122.25   
1203           0           0           2  ...         6.0      122.25   
1204           0           0           1  ...         3.6       88.61   
1205           1           0           1  ...         6.0      108.79   

     Odor_AgeAdj  PainIntens_RawScore PainInterf_Tscore  Taste_Unadj  \
1202      110.45                  0.0              38.6        84.07   
1203      111.41                  7.0              63.8       110.65   
1204       64.58                  3.0              50.1       117.16   
1205       97.19                  0.0              38.6        75.43   

      Taste_AgeAdj  Mars_Log_Score  Mars_Errs  Mars_Final  
1202         84.25            1.80        1.0        1.76  
1203        109.73            1.80        1.0        1.76  
1204        117.40            1.80        0.0        1.80  
1205         73.72            1.84        0.0        1.84  

[4 rows x 499 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1206 entries, 0 to 1205
Columns: 499 entries, Subject to Mars_Final
dtypes: float64(410), int64(25), object(64)
memory usage: 4.6+ MB


In [ ]:
df.describe()
# Standard deviation is a measure of how spread out the numbers are in a dataset.
# It tells you how much the values deviate from the mean (average).
# If the standard deviation is small, most values are close to the mean. If it's large, values are more spread out.

Subject   MMSE_Score   PSQI_Score   PSQI_Comp1   PSQI_Comp2  \
count    1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   
mean   374551.585406    28.985904     4.854892     0.903814     0.981758   
std    272686.898230     1.048516     2.802134     0.631696     0.828736   
min    100004.000000    23.000000     0.000000     0.000000     0.000000   
25%    154254.250000    28.000000     3.000000     1.000000     0.000000   
50%    212166.500000    29.000000     4.000000     1.000000     1.000000   
75%    586310.500000    30.000000     6.000000     1.000000     1.000000   
max    996782.000000    30.000000    19.000000     3.000000     3.000000   

        PSQI_Comp3   PSQI_Comp4   PSQI_Comp5   PSQI_Comp6   PSQI_Comp7  ...  \
count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000  ...   
mean      0.575456     0.466003     1.105307     0.235489     0.587065  ...   
std       0.817157     0.807694     0.510167     0.671359     0.637916  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     1.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     1.000000     0.000000     1.000000  ...   
75%       1.000000     1.000000     1.000000     0.000000     1.000000  ...   
max       3.000000     3.000000     3.000000     3.000000     3.000000  ...   

        Noise_Comp   Odor_Unadj  Odor_AgeAdj  PainIntens_RawScore  \
count  1196.000000  1204.000000  1204.000000          1201.000000   
mean      4.404682   110.421321    97.727500             1.449625   
std       1.513823     9.107963    11.273251             1.783069   
min      -0.400000    82.740000    59.860000             0.000000   
25%       3.600000   101.120000    87.110000             0.000000   
50%       4.400000   108.790000    98.040000             1.000000   
75%       5.200000   122.250000   110.450000             2.000000   
max      14.000000   122.250000   111.410000            10.000000   

       PainInterf_Tscore  Taste_Unadj  Taste_AgeAdj  Mars_Log_Score  \
count        1205.000000  1200.000000   1200.000000     1198.000000   
mean           45.847718    95.166983     93.998533        1.845467   
std             7.679288    14.583412     14.837851        0.541393   
min            38.600000    56.350000     59.500000        1.560000   
25%            38.600000    84.070000     83.220000        1.800000   
50%            45.900000    95.360000     94.970000        1.800000   
75%            52.200000   105.570000    102.920000        1.880000   
max            75.300000   134.650000    131.380000       15.000000   

         Mars_Errs   Mars_Final  
count  1195.000000  1195.000000  
mean      0.581590     1.822251  
std       0.973172     0.542893  
min       0.000000     1.080000  
25%       0.000000     1.760000  
50%       0.000000     1.800000  
75%       1.000000     1.840000  
max      17.000000    15.000000  

[8 rows x 435 columns]

In [ ]:
df.columns

Index(['Subject', 'Gender', 'Age', 'MMSE_Score', 'PSQI_Score', 'PSQI_Comp1',
       'PSQI_Comp2', 'PSQI_Comp3', 'PSQI_Comp4', 'PSQI_Comp5',
       ...
       'Noise_Comp', 'Odor_Unadj', 'Odor_AgeAdj', 'PainIntens_RawScore',
       'PainInterf_Tscore', 'Taste_Unadj', 'Taste_AgeAdj', 'Mars_Log_Score',
       'Mars_Errs', 'Mars_Final'],
      dtype='object', length=499)

In [ ]:
df['Gender'].unique()

array(['M', 'F'], dtype=object)

In [ ]:
df['Gender'].value_counts()

Gender
F    656
M    550
Name: count, dtype: int64

In [ ]:
df[df['Gender']=="F"] # Display only the rows of female subjects

Subject Gender    Age  MMSE_Score  PSQI_Score  PSQI_Comp1  PSQI_Comp2  \
2      100307      F  26-30          29           4           1           0   
5      101006      F  31-35          28           2           1           1   
7      101208      F  31-35          30           6           1           2   
10     101612      F  26-30          28           4           1           1   
11     101915      F  31-35          29           6           1           1   
...       ...    ...    ...         ...         ...         ...         ...   
1194   984472      F  26-30          30           5           1           1   
1196   987983      F  26-30          30           3           1           1   
1200   992673      F  31-35          30           0           0           0   
1202   993675      F  26-30          30           4           1           1   
1205   996782      F  26-30          30           4           1           1   

      PSQI_Comp3  PSQI_Comp4  PSQI_Comp5  ...  Noise_Comp  Odor_Unadj  \
2              1           0           2  ...         3.6      101.12   
5              0           0           0  ...         6.0      122.25   
7              0           0           2  ...         4.4      101.12   
10             0           0           1  ...         4.4      122.25   
11             1           1           1  ...         4.4       96.87   
...          ...         ...         ...  ...         ...         ...   
1194           1           0           1  ...         2.8      122.25   
1196           0           0           1  ...         5.2      108.79   
1200           0           0           0  ...         3.6      122.25   
1202           1           0           1  ...         0.4      122.25   
1205           1           0           1  ...         6.0      108.79   

     Odor_AgeAdj  PainIntens_RawScore PainInterf_Tscore  Taste_Unadj  \
2          86.45                  0.0              38.6        71.69   
5         111.41                  0.0              38.6       123.80   
7          87.11                  1.0              50.1       105.57   
10        111.41                  2.0              48.7        97.26   
11         77.61                  0.0              38.6       112.11   
...          ...                  ...               ...          ...   
1194      110.45                  0.0              38.6       108.73   
1196       97.19                  5.0              56.4        88.02   
1200      111.41                  1.0              38.6       101.63   
1202      110.45                  0.0              38.6        84.07   
1205       97.19                  0.0              38.6        75.43   

      Taste_AgeAdj  Mars_Log_Score  Mars_Errs  Mars_Final  
2            71.76            1.76        0.0        1.76  
5           123.31            1.80        0.0        1.80  
7           102.32            1.92        0.0        1.92  
10           96.41            1.84        0.0        1.84  
11          111.70            1.84        1.0        1.80  
...            ...             ...        ...         ...  
1194        108.00            1.76        0.0        1.76  
1196         87.70            1.88        0.0        1.88  
1200         99.26            1.80        0.0        1.80  
1202         84.25            1.80        1.0        1.76  
1205         73.72            1.84        0.0        1.84  

[656 rows x 499 columns]

### 1.2 Documentation
Woah! So many columns and abbreviations! What do they all mean? Make sure you know where your [dataset's documentation](https://www.humanconnectome.org/storage/app/media/documentation/s1200/HCP_S1200_Release_Reference_Manual.pdf) is.

https://wiki.humanconnectome.org/#resources-for-public-hcp-data-users

Unfortunately, thorough documentation is not always available. Some data types are also very field-specific and require the help of experts, which is part of what makes machine learning so wonderfully interdisciplinary.

In [ ]:
list(df.columns)

['Subject',
 'Gender',
 'Age',
 'MMSE_Score',
 'PSQI_Score',
 'PSQI_Comp1',
 'PSQI_Comp2',
 'PSQI_Comp3',
 'PSQI_Comp4',
 'PSQI_Comp5',
 'PSQI_Comp6',
 'PSQI_Comp7',
 'PSQI_BedTime',
 'PSQI_Min2Asleep',
 'PSQI_GetUpTime',
 'PSQI_AmtSleep',
 'PSQI_Latency30Min',
 'PSQI_WakeUp',
 'PSQI_Bathroom',
 'PSQI_Breathe',
 'PSQI_Snore',
 'PSQI_TooCold',
 'PSQI_TooHot',
 'PSQI_BadDream',
 'PSQI_Pain',
 'PSQI_Other',
 'PSQI_Quality',
 'PSQI_SleepMeds',
 'PSQI_DayStayAwake',
 'PSQI_DayEnthusiasm',
 'PSQI_BedPtnrRmate',
 'PicSeq_Unadj',
 'PicSeq_AgeAdj',
 'CardSort_Unadj',
 'CardSort_AgeAdj',
 'Flanker_Unadj',
 'Flanker_AgeAdj',
 'PMAT24_A_CR',
 'PMAT24_A_SI',
 'PMAT24_A_RTCR',
 'ReadEng_Unadj',
 'ReadEng_AgeAdj',
 'PicVocab_Unadj',
 'PicVocab_AgeAdj',
 'ProcSpeed_Unadj',
 'ProcSpeed_AgeAdj',
 'DDisc_SV_1mo_200',
 'DDisc_SV_6mo_200',
 'DDisc_SV_1yr_200',
 'DDisc_SV_3yr_200',
 'DDisc_SV_5yr_200',
 'DDisc_SV_10yr_200',
 'DDisc_SV_1mo_40K',
 'DDisc_SV_6mo_40K',
 'DDisc_SV_1yr_40K',
 'DDisc_SV_3yr_40K',


Handpicking the columns you want to work with:

In [ ]:
basics = ['Subject','Gender','Age','PSQI_BedTime']
df[basics]

Subject Gender    Age PSQI_BedTime
0      100004      M  22-25      9:00:00
1      100206      M  26-30     22:30:00
2      100307      F  26-30     22:00:00
3      100408      M  31-35     22:00:00
4      100610      M  26-30     21:30:00
...       ...    ...    ...          ...
1201   992774      M  31-35     22:00:00
1202   993675      F  26-30     22:00:00
1203   994273      M  26-30      2:00:00
1204   995174      M  22-25     23:00:00
1205   996782      F  26-30     23:30:00

[1206 rows x 4 columns]

Let's also add all of some cognitive variables to the mix! Specifically, we'll select the measures related to fluid intelligence (they start with `PMAT` for Penn Matrix Test) and impulsivity (they start with `DDisc` for Delay Discounting)

In [ ]:
cognition = ['Subject','Gender','Age','PSQI_BedTime']
for col in df.columns:
    if (col.find("PMAT")!=-1 or col.find("DDisc")!=-1):
        cognition.append(col)
print(f"List of variables we will be looking at: {cognition}") # PS: f-strings will be very useful for you in your Python journey!

List of variables we will be looking at: ['Subject', 'Gender', 'Age', 'PSQI_BedTime', 'PMAT24_A_CR', 'PMAT24_A_SI', 'PMAT24_A_RTCR', 'DDisc_SV_1mo_200', 'DDisc_SV_6mo_200', 'DDisc_SV_1yr_200', 'DDisc_SV_3yr_200', 'DDisc_SV_5yr_200', 'DDisc_SV_10yr_200', 'DDisc_SV_1mo_40K', 'DDisc_SV_6mo_40K', 'DDisc_SV_1yr_40K', 'DDisc_SV_3yr_40K', 'DDisc_SV_5yr_40K', 'DDisc_SV_10yr_40K', 'DDisc_AUC_200', 'DDisc_AUC_200.1', 'DDisc_AUC_40K']


Okay! So let's actually select this subset of our data, and make it a separate dataframe!

In [ ]:
df_cognition = df[cognition]

In [ ]:
df_cognition.head()

Subject Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0   100004      M  22-25      9:00:00         19.0          0.0   
1   100206      M  26-30     22:30:00         20.0          0.0   
2   100307      F  26-30     22:00:00         17.0          2.0   
3   100408      M  31-35     22:00:00          7.0         12.0   
4   100610      M  26-30     21:30:00         23.0          0.0   

   PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  ...  \
0        15590.0            153.13             46.88             21.88  ...   
1        18574.5             78.13             34.38              9.38  ...   
2        11839.0            103.13             46.88            103.13  ...   
3         3042.0            153.13             46.88             46.88  ...   
4        12280.0            196.88            196.88            184.38  ...   

   DDisc_SV_10yr_200  DDisc_SV_1mo_40K  DDisc_SV_6mo_40K  DDisc_SV_1yr_40K  \
0              21.88           34375.0           24375.0             625.0   
1               9.38           30625.0             625.0             625.0   
2               9.38           19375.0           29375.0           24375.0   
3               9.38           39375.0           29375.0           24375.0   
4             146.88           39375.0           39375.0           39375.0   

   DDisc_SV_3yr_40K  DDisc_SV_5yr_40K  DDisc_SV_10yr_40K  DDisc_AUC_200  \
0             625.0             625.0              625.0       0.121811   
1            3125.0             625.0              625.0       0.097072   
2            9375.0            9375.0             9375.0       0.162176   
3           19375.0           18125.0             4375.0       0.203061   
4           39375.0           36875.0            24375.0       0.801629   

   DDisc_AUC_200.1  DDisc_AUC_40K  
0         0.121811       0.067448  
1         0.097072       0.050000  
2         0.162176       0.311459  
3         0.203061       0.421354  
4         0.801629       0.868750  

[5 rows x 22 columns]

In [ ]:
df_cognition.shape

(1206, 22)

### 2. Merging two dataframes

So for our dataset of 1206 subjects, we have information about their gender, age-range and a variety of cognitive measures. It would be cool to be able to integrate some other data as well. You have been provided with a separate file that contains brain structure volume data obtained from the neuroimaging data of these same subjects (note that not all subjects in the HCP have neuroimaging data)

In [ ]:
df_volumes = pd.read_csv("data_csv/HCP_volumes.csv")

In [ ]:
df_volumes

Subject      TBV  Gp_Left  Gp_Right  Str_Left  Str_Right  Thal_Left  \
0      100206  1694580  1746.90   1597.69  11848.20   12030.00    7827.26   
1      100307  1319380  1489.99   1291.05   9470.57    9749.09    5743.19   
2      100408  1563630  1617.93   1531.84  10274.90   10693.40    6865.83   
3      100610  1585390  1841.22   1639.88  12464.30   12575.80    7452.36   
4      101006  1261040  1679.33   1571.28   9774.47    9967.58    5856.38   
...       ...      ...      ...       ...       ...        ...        ...   
1081   992774  1328640  1401.50   1266.36   9735.71    9799.51    6253.92   
1082   993675  1257810  1394.98   1326.72   9093.62    9385.51    5994.61   
1083   994273  1487180  1646.74   1501.31  11164.60   11670.90    6711.82   
1084   995174  1517230  1678.30   1594.26  11154.40   11338.60    6766.36   
1085   996782  1555360  1560.65   1477.30  11323.50   11589.60    7095.64   

      Thal_Right  str_left_SA  str_right_SA  thal_left_SA  thal_right_SA  \
0        7750.77  15156.32547   16506.40087   7902.211643    8394.220729   
1        5773.38  12555.09774   13845.51983   6368.274739    6969.617635   
2        6842.16  14165.48884   15590.63180   7235.193241    7763.609633   
3        7598.48  15493.70123   16940.05756   7594.851693    8316.436428   
4        5877.30  13309.15216   14458.50988   6514.710165    7003.367310   
...          ...          ...           ...           ...            ...   
1081     6234.37  13763.48378   14650.11549   6814.085943    7170.076369   
1082     5971.29  12477.92590   13833.70956   6601.958682    7131.259938   
1083     6886.75  14637.99165   16242.63411   7100.454111    7637.944904   
1084     6734.46  14516.88893   15906.36860   7263.501869    7809.016115   
1085     7137.14  14794.60156   15931.95362   7368.394110    7834.162855   

       gp_left_SA  gp_right_SA  
0     3603.322526  3751.440646  
1     3149.392096  3156.322277  
2     3476.901847  3725.802635  
3     3862.828117  3893.266274  
4     3586.431707  3773.869259  
...           ...          ...  
1081  3198.083973  3262.912737  
1082  2995.664299  3186.251544  
1083  3441.358390  3549.415428  
1084  3535.013585  3761.052946  
1085  3319.287200  3451.372013  

[1086 rows x 14 columns]

In [ ]:
df_volumes['TBV'].mean() # mean Total Brain Volume!

np.float64(1406181.3996316758)

In [ ]:
df_volumes['TBV'].std() # standard deviation of TBV!

151214.5002924137

List all our subjects whose TBV is above average! (PS: another way to write this: `df_volumes[df_volumes['TBV']>1406181]`)

In [ ]:
df_volumes[df_volumes['TBV']>df_volumes['TBV'].mean()]

Subject      TBV  Gp_Left  Gp_Right  Str_Left  Str_Right  Thal_Left  \
0      100206  1694580  1746.90   1597.69  11848.20   12030.00    7827.26   
2      100408  1563630  1617.93   1531.84  10274.90   10693.40    6865.83   
3      100610  1585390  1841.22   1639.88  12464.30   12575.80    7452.36   
6      101309  1426200  1690.65   1507.14  10144.90   10251.60    6266.95   
7      101410  1588370  1440.26   1328.78   9376.25    9713.07    6865.83   
...       ...      ...      ...       ...       ...        ...        ...   
1075   987074  1451930  1541.44   1384.69  10053.70   10266.00    6290.62   
1079   991267  1802370  1804.52   1681.04  12363.10   12466.30    7433.50   
1083   994273  1487180  1646.74   1501.31  11164.60   11670.90    6711.82   
1084   995174  1517230  1678.30   1594.26  11154.40   11338.60    6766.36   
1085   996782  1555360  1560.65   1477.30  11323.50   11589.60    7095.64   

      Thal_Right  str_left_SA  str_right_SA  thal_left_SA  thal_right_SA  \
0        7750.77  15156.32547   16506.40087   7902.211643    8394.220729   
2        6842.16  14165.48884   15590.63180   7235.193241    7763.609633   
3        7598.48  15493.70123   16940.05756   7594.851693    8316.436428   
6        6182.92  13848.22982   15324.77410   6798.623361    7329.928997   
7        6683.70  13217.76480   14592.03987   7153.875176    7624.042048   
...          ...          ...           ...           ...            ...   
1075     6012.10  13970.55221   14972.19121   7221.456229    7054.193554   
1079     7367.64  15554.95026   16965.02114   7649.207918    8214.273885   
1083     6886.75  14637.99165   16242.63411   7100.454111    7637.944904   
1084     6734.46  14516.88893   15906.36860   7263.501869    7809.016115   
1085     7137.14  14794.60156   15931.95362   7368.394110    7834.162855   

       gp_left_SA  gp_right_SA  
0     3603.322526  3751.440646  
2     3476.901847  3725.802635  
3     3862.828117  3893.266274  
6     3582.231784  3677.827998  
7     3143.810997  3339.830347  
...           ...          ...  
1075  3353.157608  3502.805296  
1079  3620.911699  3828.376050  
1083  3441.358390  3549.415428  
1084  3535.013585  3761.052946  
1085  3319.287200  3451.372013  

[509 rows x 14 columns]

Ok! Let's merge our dataframes. Do they have the same number of subjects?

In [ ]:
print(df_volumes.shape)
print(df_cognition.shape)

(1086, 14)
(1206, 22)


No! We only have volume data for 1086 subjects... So how do we create a new dataframe, where we only keep the subjects for which we have all of our features?

You need a common column in both datasets to merge on (e.g., 'Subject', 'ID', etc.) -> this is called the **key merge**

In [ ]:
df_final=pd.merge(df_cognition,df_volumes, on='Subject')
df_final.head()

Subject Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0   100206      M  26-30     22:30:00         20.0          0.0   
1   100307      F  26-30     22:00:00         17.0          2.0   
2   100408      M  31-35     22:00:00          7.0         12.0   
3   100610      M  26-30     21:30:00         23.0          0.0   
4   101006      F  31-35     23:00:00         11.0          8.0   

   PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  ...  \
0        18574.5             78.13             34.38              9.38  ...   
1        11839.0            103.13             46.88            103.13  ...   
2         3042.0            153.13             46.88             46.88  ...   
3        12280.0            196.88            196.88            184.38  ...   
4         6569.0            140.63             96.88            115.63  ...   

   Str_Left  Str_Right  Thal_Left  Thal_Right  str_left_SA  str_right_SA  \
0  11848.20   12030.00    7827.26     7750.77  15156.32547   16506.40087   
1   9470.57    9749.09    5743.19     5773.38  12555.09774   13845.51983   
2  10274.90   10693.40    6865.83     6842.16  14165.48884   15590.63180   
3  12464.30   12575.80    7452.36     7598.48  15493.70123   16940.05756   
4   9774.47    9967.58    5856.38     5877.30  13309.15216   14458.50988   

   thal_left_SA  thal_right_SA   gp_left_SA  gp_right_SA  
0   7902.211643    8394.220729  3603.322526  3751.440646  
1   6368.274739    6969.617635  3149.392096  3156.322277  
2   7235.193241    7763.609633  3476.901847  3725.802635  
3   7594.851693    8316.436428  3862.828117  3893.266274  
4   6514.710165    7003.367310  3586.431707  3773.869259  

[5 rows x 35 columns]

In [ ]:
df_final.shape

(1086, 35)

Mission accomplished! We have 1086 and 14 + 22 features.

Merge the Datasets

💡 Option A: Inner join (only keep rows that match in both) -> intersection

merged = pd.merge(df1, df2, on='Subject', how='inner')

💡 Option B: Left join (keep all rows from df1)

merged = pd.merge(df1, df2, on='Subject', how='left')

💡 Option C: Outer join (keep all rows from both, fill missing with NaN) -> union

merged = pd.merge(df1, df2, on='Subject', how='outer')

❓What If There's No Common Column?

Then you can't merge based on identity — you might want to use:

pd.concat([df1, df2], axis=1) → if you want to combine columns side-by-side

pd.concat([df1, df2], axis=0) → if they have same columns and you want to stack rows

## 3. Removing features that you do not need

In [ ]:
df_final.columns

Index(['Subject', 'Gender', 'Age', 'PSQI_BedTime', 'PMAT24_A_CR',
       'PMAT24_A_SI', 'PMAT24_A_RTCR', 'DDisc_SV_1mo_200', 'DDisc_SV_6mo_200',
       'DDisc_SV_1yr_200', 'DDisc_SV_3yr_200', 'DDisc_SV_5yr_200',
       'DDisc_SV_10yr_200', 'DDisc_SV_1mo_40K', 'DDisc_SV_6mo_40K',
       'DDisc_SV_1yr_40K', 'DDisc_SV_3yr_40K', 'DDisc_SV_5yr_40K',
       'DDisc_SV_10yr_40K', 'DDisc_AUC_200', 'DDisc_AUC_200.1',
       'DDisc_AUC_40K', 'TBV', 'Gp_Left', 'Gp_Right', 'Str_Left', 'Str_Right',
       'Thal_Left', 'Thal_Right', 'str_left_SA', 'str_right_SA',
       'thal_left_SA', 'thal_right_SA', 'gp_left_SA', 'gp_right_SA'],
      dtype='object')

Do we have any columns that repeat themselves? Yes we do! `DDisc_AUC_200` and `DDisc_AUC_200.1` might be the same. Let's check

In [ ]:
df_final["DDisc_AUC_200"].equals(df_final["DDisc_AUC_200.1"])

True

Nice! Let's remove column `DDisc_AUC_200.1` then.

In [ ]:
df_final.drop('DDisc_AUC_200.1', inplace=True, axis=1)

axis=1: means you're dropping a column (not a row)

axis=0 would mean a row

inplace=True: tells pandas to change df_final directly

If False, you'd need to assign the result like:

df_final = df_final.drop(...)

In [ ]:
df_final.columns

Index(['Subject', 'Gender', 'Age', 'PSQI_BedTime', 'PMAT24_A_CR',
       'PMAT24_A_SI', 'PMAT24_A_RTCR', 'DDisc_SV_1mo_200', 'DDisc_SV_6mo_200',
       'DDisc_SV_1yr_200', 'DDisc_SV_3yr_200', 'DDisc_SV_5yr_200',
       'DDisc_SV_10yr_200', 'DDisc_SV_1mo_40K', 'DDisc_SV_6mo_40K',
       'DDisc_SV_1yr_40K', 'DDisc_SV_3yr_40K', 'DDisc_SV_5yr_40K',
       'DDisc_SV_10yr_40K', 'DDisc_AUC_200', 'DDisc_AUC_40K', 'TBV', 'Gp_Left',
       'Gp_Right', 'Str_Left', 'Str_Right', 'Thal_Left', 'Thal_Right',
       'str_left_SA', 'str_right_SA', 'thal_left_SA', 'thal_right_SA',
       'gp_left_SA', 'gp_right_SA'],
      dtype='object')

## 4. Making your data machine-readable

To be machine-readable, your variables need to be numerical. Want to check?

In [ ]:
df_final.dtypes

Subject                int64
Gender                object
Age                   object
PSQI_BedTime          object
PMAT24_A_CR          float64
PMAT24_A_SI          float64
PMAT24_A_RTCR        float64
DDisc_SV_1mo_200     float64
DDisc_SV_6mo_200     float64
DDisc_SV_1yr_200     float64
DDisc_SV_3yr_200     float64
DDisc_SV_5yr_200     float64
DDisc_SV_10yr_200    float64
DDisc_SV_1mo_40K     float64
DDisc_SV_6mo_40K     float64
DDisc_SV_1yr_40K     float64
DDisc_SV_3yr_40K     float64
DDisc_SV_5yr_40K     float64
DDisc_SV_10yr_40K    float64
DDisc_AUC_200        float64
DDisc_AUC_40K        float64
TBV                    int64
Gp_Left              float64
Gp_Right             float64
Str_Left             float64
Str_Right            float64
Thal_Left            float64
Thal_Right           float64
str_left_SA          float64
str_right_SA         float64
thal_left_SA         float64
thal_right_SA        float64
gp_left_SA           float64
gp_right_SA          float64
dtype: object

We have 3 columns that are non-numerical: `Gender`,`Age`,`PSQI_BedTime`. Let's figure out how to handle them, one at a time. First of all, we know that the `Gender` column is categorical:

In [ ]:
# "value_counts" say how many times each unique value appears in the Gender column of your DataFrame.
df_final['Gender'].value_counts()

Gender
F    590
M    496
Name: count, dtype: int64

### 4.1 One-hot encoding or binarizing your data
In this specific dataset, the `Gender` column has one of two values: `M` or `F`. Given that we only have two categories here, you can just replace all of your `M` values with 1 and `F` values with 2.

Indeed, it has one of two values: `M` or `F`. Given that we only have two categories here, you can easily just set all of the `M`s to 1 and the `F` to 2:

In [ ]:
df_final['Gender'] = df_final['Gender'].replace('M',1)
df_final['Gender'] = df_final['Gender'].replace('F',2)
df_final

<ipython-input-84-7e813a8156c1>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_final['Gender'] = df_final['Gender'].replace('F',2)


Subject  Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0      100206       1  26-30     22:30:00         20.0          0.0   
1      100307       2  26-30     22:00:00         17.0          2.0   
2      100408       1  31-35     22:00:00          7.0         12.0   
3      100610       1  26-30     21:30:00         23.0          0.0   
4      101006       2  31-35     23:00:00         11.0          8.0   
...       ...     ...    ...          ...          ...          ...   
1081   992774       1  31-35     22:00:00         14.0          5.0   
1082   993675       2  26-30     22:00:00         21.0          0.0   
1083   994273       1  26-30      2:00:00         20.0          0.0   
1084   995174       1  22-25     23:00:00         19.0          0.0   
1085   996782       2  26-30     23:30:00         20.0          0.0   

      PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  \
0           18574.5             78.13             34.38              9.38   
1           11839.0            103.13             46.88            103.13   
2            3042.0            153.13             46.88             46.88   
3           12280.0            196.88            196.88            184.38   
4            6569.0            140.63             96.88            115.63   
...             ...               ...               ...               ...   
1081        15334.5              3.13              9.38              3.13   
1082        22151.0            159.38            165.63            146.88   
1083        33421.0            153.13             21.88              3.13   
1084        18161.0            190.63            103.13            121.88   
1085        18360.0            196.88            159.38            159.38   

      ...  Str_Left  Str_Right  Thal_Left  Thal_Right  str_left_SA  \
0     ...  11848.20   12030.00    7827.26     7750.77  15156.32547   
1     ...   9470.57    9749.09    5743.19     5773.38  12555.09774   
2     ...  10274.90   10693.40    6865.83     6842.16  14165.48884   
3     ...  12464.30   12575.80    7452.36     7598.48  15493.70123   
4     ...   9774.47    9967.58    5856.38     5877.30  13309.15216   
...   ...       ...        ...        ...         ...          ...   
1081  ...   9735.71    9799.51    6253.92     6234.37  13763.48378   
1082  ...   9093.62    9385.51    5994.61     5971.29  12477.92590   
1083  ...  11164.60   11670.90    6711.82     6886.75  14637.99165   
1084  ...  11154.40   11338.60    6766.36     6734.46  14516.88893   
1085  ...  11323.50   11589.60    7095.64     7137.14  14794.60156   

      str_right_SA  thal_left_SA  thal_right_SA   gp_left_SA  gp_right_SA  
0      16506.40087   7902.211643    8394.220729  3603.322526  3751.440646  
1      13845.51983   6368.274739    6969.617635  3149.392096  3156.322277  
2      15590.63180   7235.193241    7763.609633  3476.901847  3725.802635  
3      16940.05756   7594.851693    8316.436428  3862.828117  3893.266274  
4      14458.50988   6514.710165    7003.367310  3586.431707  3773.869259  
...            ...           ...            ...          ...          ...  
1081   14650.11549   6814.085943    7170.076369  3198.083973  3262.912737  
1082   13833.70956   6601.958682    7131.259938  2995.664299  3186.251544  
1083   16242.63411   7100.454111    7637.944904  3441.358390  3549.415428  
1084   15906.36860   7263.501869    7809.016115  3535.013585  3761.052946  
1085   15931.95362   7368.394110    7834.162855  3319.287200  3451.372013  

[1086 rows x 34 columns]

However, suppose that you actually had more than 2 numerical values for this feature (e.g. `M`,`F`,`other`). If you just convert categorical variables to numerical values (ex: `M`=1,`F`=2,`other`=3), you give a "distance" to the relationship between variables. For instance, since 1 is closer to 2 than to 3, you are telling your machine that `M` is "closer" to `F` (`distance = 2 - 1 = 1`) than to `other` (`distance = 3 - 1 = 2`). We want our categories to be independent. That's where one hot encoding comes into play: "[A representation of categorical variables as binary vectors](https://machinelearningmastery.com/how-to-one-hot-encode-sequence-data-in-python/)"\

Let's revert our `Gender` feature back to its original form, and see how to one-hot encode it.

**One-hot encoding** is a way to convert categorical data (like "Male", "Female", "Other") into numerical form so machine learning models can understand it. -> **without imposing non-existent order or distance between categories**.

One-hot encoding fixes this by **creating a new binary column for each category.**

In [ ]:
df_final['Gender'] = df_final['Gender'].replace(1,'M')
df_final['Gender'] = df_final['Gender'].replace(2,'F')

one_hot = pd.get_dummies(df_final['Gender'])
one_hot

F      M
0     False   True
1      True  False
2     False   True
3     False   True
4      True  False
...     ...    ...
1081  False   True
1082   True  False
1083  False   True
1084  False   True
1085   True  False

[1086 rows x 2 columns]

As you can see, we convert 1 categorical feature into N binary features, where N = number of values that your feature can take on (2 in this example). Let's add it to our dataframe!

In [ ]:
df_final.join(one_hot)

Subject Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0      100206      M  26-30     22:30:00         20.0          0.0   
1      100307      F  26-30     22:00:00         17.0          2.0   
2      100408      M  31-35     22:00:00          7.0         12.0   
3      100610      M  26-30     21:30:00         23.0          0.0   
4      101006      F  31-35     23:00:00         11.0          8.0   
...       ...    ...    ...          ...          ...          ...   
1081   992774      M  31-35     22:00:00         14.0          5.0   
1082   993675      F  26-30     22:00:00         21.0          0.0   
1083   994273      M  26-30      2:00:00         20.0          0.0   
1084   995174      M  22-25     23:00:00         19.0          0.0   
1085   996782      F  26-30     23:30:00         20.0          0.0   

      PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  \
0           18574.5             78.13             34.38              9.38   
1           11839.0            103.13             46.88            103.13   
2            3042.0            153.13             46.88             46.88   
3           12280.0            196.88            196.88            184.38   
4            6569.0            140.63             96.88            115.63   
...             ...               ...               ...               ...   
1081        15334.5              3.13              9.38              3.13   
1082        22151.0            159.38            165.63            146.88   
1083        33421.0            153.13             21.88              3.13   
1084        18161.0            190.63            103.13            121.88   
1085        18360.0            196.88            159.38            159.38   

      ...  Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0     ...    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1     ...    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2     ...    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3     ...    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4     ...    5856.38     5877.30  13309.15216   14458.50988   6514.710165   
...   ...        ...         ...          ...           ...           ...   
1081  ...    6253.92     6234.37  13763.48378   14650.11549   6814.085943   
1082  ...    5994.61     5971.29  12477.92590   13833.70956   6601.958682   
1083  ...    6711.82     6886.75  14637.99165   16242.63411   7100.454111   
1084  ...    6766.36     6734.46  14516.88893   15906.36860   7263.501869   
1085  ...    7095.64     7137.14  14794.60156   15931.95362   7368.394110   

      thal_right_SA   gp_left_SA  gp_right_SA      F      M  
0       8394.220729  3603.322526  3751.440646  False   True  
1       6969.617635  3149.392096  3156.322277   True  False  
2       7763.609633  3476.901847  3725.802635  False   True  
3       8316.436428  3862.828117  3893.266274  False   True  
4       7003.367310  3586.431707  3773.869259   True  False  
...             ...          ...          ...    ...    ...  
1081    7170.076369  3198.083973  3262.912737  False   True  
1082    7131.259938  2995.664299  3186.251544   True  False  
1083    7637.944904  3441.358390  3549.415428  False   True  
1084    7809.016115  3535.013585  3761.052946  False   True  
1085    7834.162855  3319.287200  3451.372013   True  False  

[1086 rows x 36 columns]

In [ ]:
df_final.head()

Subject Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0   100206      M  26-30     22:30:00         20.0          0.0   
1   100307      F  26-30     22:00:00         17.0          2.0   
2   100408      M  31-35     22:00:00          7.0         12.0   
3   100610      M  26-30     21:30:00         23.0          0.0   
4   101006      F  31-35     23:00:00         11.0          8.0   

   PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  ...  \
0        18574.5             78.13             34.38              9.38  ...   
1        11839.0            103.13             46.88            103.13  ...   
2         3042.0            153.13             46.88             46.88  ...   
3        12280.0            196.88            196.88            184.38  ...   
4         6569.0            140.63             96.88            115.63  ...   

   Str_Left  Str_Right  Thal_Left  Thal_Right  str_left_SA  str_right_SA  \
0  11848.20   12030.00    7827.26     7750.77  15156.32547   16506.40087   
1   9470.57    9749.09    5743.19     5773.38  12555.09774   13845.51983   
2  10274.90   10693.40    6865.83     6842.16  14165.48884   15590.63180   
3  12464.30   12575.80    7452.36     7598.48  15493.70123   16940.05756   
4   9774.47    9967.58    5856.38     5877.30  13309.15216   14458.50988   

   thal_left_SA  thal_right_SA   gp_left_SA  gp_right_SA  
0   7902.211643    8394.220729  3603.322526  3751.440646  
1   6368.274739    6969.617635  3149.392096  3156.322277  
2   7235.193241    7763.609633  3476.901847  3725.802635  
3   7594.851693    8316.436428  3862.828117  3893.266274  
4   6514.710165    7003.367310  3586.431707  3773.869259  

[5 rows x 34 columns]

Make sure you reassign the modification you made to `df_final`, or it will not save!

In [ ]:
df_final = df_final.join(one_hot)
df_final.head()

Subject Gender    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  \
0   100206      M  26-30     22:30:00         20.0          0.0   
1   100307      F  26-30     22:00:00         17.0          2.0   
2   100408      M  31-35     22:00:00          7.0         12.0   
3   100610      M  26-30     21:30:00         23.0          0.0   
4   101006      F  31-35     23:00:00         11.0          8.0   

   PMAT24_A_RTCR  DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  ...  \
0        18574.5             78.13             34.38              9.38  ...   
1        11839.0            103.13             46.88            103.13  ...   
2         3042.0            153.13             46.88             46.88  ...   
3        12280.0            196.88            196.88            184.38  ...   
4         6569.0            140.63             96.88            115.63  ...   

   Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4    5856.38     5877.30  13309.15216   14458.50988   6514.710165   

   thal_right_SA   gp_left_SA  gp_right_SA      F      M  
0    8394.220729  3603.322526  3751.440646  False   True  
1    6969.617635  3149.392096  3156.322277   True  False  
2    7763.609633  3476.901847  3725.802635  False   True  
3    8316.436428  3862.828117  3893.266274  False   True  
4    7003.367310  3586.431707  3773.869259   True  False  

[5 rows x 36 columns]

In [ ]:
df_final = df_final.drop('Gender',axis = 1)  # Drop column Gender as it is now one-hot encoded
df_final.head()

Subject    Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  PMAT24_A_RTCR  \
0   100206  26-30     22:30:00         20.0          0.0        18574.5   
1   100307  26-30     22:00:00         17.0          2.0        11839.0   
2   100408  31-35     22:00:00          7.0         12.0         3042.0   
3   100610  26-30     21:30:00         23.0          0.0        12280.0   
4   101006  31-35     23:00:00         11.0          8.0         6569.0   

   DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  DDisc_SV_3yr_200  \
0             78.13             34.38              9.38             21.88   
1            103.13             46.88            103.13             21.88   
2            153.13             46.88             46.88             46.88   
3            196.88            196.88            184.38            171.88   
4            140.63             96.88            115.63             46.88   

   ...  Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0  ...    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1  ...    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2  ...    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3  ...    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4  ...    5856.38     5877.30  13309.15216   14458.50988   6514.710165   

   thal_right_SA   gp_left_SA  gp_right_SA      F      M  
0    8394.220729  3603.322526  3751.440646  False   True  
1    6969.617635  3149.392096  3156.322277   True  False  
2    7763.609633  3476.901847  3725.802635  False   True  
3    8316.436428  3862.828117  3893.266274  False   True  
4    7003.367310  3586.431707  3773.869259   True  False  

[5 rows x 35 columns]

Now, several people have been wondering... What happens if you have a categorical variables with a LOT of different possible values (e.g. country data). Wouldn't one-hot encoding induce the problem of sparsity (i.e. a matrix with too many 0s and 1s)? Indeed, this could be an issue. Kate Nimegeers however had a really creative idea: one could theoretically replace the "Country" variables with 2 variables for the latitude and longitude of the centroid of each country in the dataset (or one could find a way to create some "aggregate measure" of latitude and longitude). Thus, there is lots of room for creativity :)

### 4.2 Parsing strings in your data

#### 4.2.1 Handling the `Age` feature

Recall: our `Age` feature is also non-numerical! Luckily, we have an age range, so the string can be split up into two new columns: min age and max age. Let's replace our age range in `df_final` with an average age estimate for each subject.

In [ ]:
df_final['Age'].value_counts()

Age
26-30    476
31-35    371
22-25    226
36+       13
Name: count, dtype: int64

For the sake of this exercise, let's replace 36+ with a range of 36-40.

In [ ]:
df_final['Age'] = df_final['Age'].replace("36+",'36-40')
df_final['Age'].value_counts()

Age
26-30    476
31-35    371
22-25    226
36-40     13
Name: count, dtype: int64

In [ ]:
fix_age = df_final['Age'].str.split('-', n=1, expand=True)
fix_age.columns = ['min','max']
fix_age
# fix_age['mean'] = (fix_age['max']+fix_age['min'])/2

min max
0     26  30
1     26  30
2     31  35
3     26  30
4     31  35
...   ..  ..
1081  31  35
1082  26  30
1083  26  30
1084  22  25
1085  26  30

[1086 rows x 2 columns]

In [ ]:
fix_age["min"] = fix_age['min'].astype(float)
fix_age["max"] = fix_age['max'].astype(float)
fix_age['mean'] = (fix_age['max']+fix_age['min'])/2
fix_age

min   max  mean
0     26.0  30.0  28.0
1     26.0  30.0  28.0
2     31.0  35.0  33.0
3     26.0  30.0  28.0
4     31.0  35.0  33.0
...    ...   ...   ...
1081  31.0  35.0  33.0
1082  26.0  30.0  28.0
1083  26.0  30.0  28.0
1084  22.0  25.0  23.5
1085  26.0  30.0  28.0

[1086 rows x 3 columns]

In [ ]:
df_final['Age'] = fix_age['mean']
df_final

Subject   Age PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  PMAT24_A_RTCR  \
0      100206  28.0     22:30:00         20.0          0.0        18574.5   
1      100307  28.0     22:00:00         17.0          2.0        11839.0   
2      100408  33.0     22:00:00          7.0         12.0         3042.0   
3      100610  28.0     21:30:00         23.0          0.0        12280.0   
4      101006  33.0     23:00:00         11.0          8.0         6569.0   
...       ...   ...          ...          ...          ...            ...   
1081   992774  33.0     22:00:00         14.0          5.0        15334.5   
1082   993675  28.0     22:00:00         21.0          0.0        22151.0   
1083   994273  28.0      2:00:00         20.0          0.0        33421.0   
1084   995174  23.5     23:00:00         19.0          0.0        18161.0   
1085   996782  28.0     23:30:00         20.0          0.0        18360.0   

      DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  DDisc_SV_3yr_200  \
0                78.13             34.38              9.38             21.88   
1               103.13             46.88            103.13             21.88   
2               153.13             46.88             46.88             46.88   
3               196.88            196.88            184.38            171.88   
4               140.63             96.88            115.63             46.88   
...                ...               ...               ...               ...   
1081              3.13              9.38              3.13              3.13   
1082            159.38            165.63            146.88             96.88   
1083            153.13             21.88              3.13             21.88   
1084            190.63            103.13            121.88             90.63   
1085            196.88            159.38            159.38             96.88   

      ...  Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0     ...    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1     ...    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2     ...    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3     ...    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4     ...    5856.38     5877.30  13309.15216   14458.50988   6514.710165   
...   ...        ...         ...          ...           ...           ...   
1081  ...    6253.92     6234.37  13763.48378   14650.11549   6814.085943   
1082  ...    5994.61     5971.29  12477.92590   13833.70956   6601.958682   
1083  ...    6711.82     6886.75  14637.99165   16242.63411   7100.454111   
1084  ...    6766.36     6734.46  14516.88893   15906.36860   7263.501869   
1085  ...    7095.64     7137.14  14794.60156   15931.95362   7368.394110   

      thal_right_SA   gp_left_SA  gp_right_SA      F      M  
0       8394.220729  3603.322526  3751.440646  False   True  
1       6969.617635  3149.392096  3156.322277   True  False  
2       7763.609633  3476.901847  3725.802635  False   True  
3       8316.436428  3862.828117  3893.266274  False   True  
4       7003.367310  3586.431707  3773.869259   True  False  
...             ...          ...          ...    ...    ...  
1081    7170.076369  3198.083973  3262.912737  False   True  
1082    7131.259938  2995.664299  3186.251544   True  False  
1083    7637.944904  3441.358390  3549.415428  False   True  
1084    7809.016115  3535.013585  3761.052946  False   True  
1085    7834.162855  3319.287200  3451.372013   True  False  

[1086 rows x 35 columns]

#### 4.2.2 Handling the `PSQI_BedTime` feature

Convert your bed time variable from HH:MM:SS to seconds!

In [ ]:
df_final['PSQI_BedTime']

0       22:30:00
1       22:00:00
2       22:00:00
3       21:30:00
4       23:00:00
          ...   
1081    22:00:00
1082    22:00:00
1083     2:00:00
1084    23:00:00
1085    23:30:00
Name: PSQI_BedTime, Length: 1086, dtype: object

In [ ]:
# "to_timedelta" Converts time strings like '23:15:00' into timedelta objects
# .dt.total_seconds() gives the number of seconds from 00:00:00

df_final['PSQI_BedTime'] = pd.to_timedelta(df_final['PSQI_BedTime']).dt.total_seconds()

In [ ]:
ftr = [3600,60,1]
for i in range(len(df_final['PSQI_BedTime'])):
    x = sum([a*b for a,b in zip(ftr, map(int,df_final['PSQI_BedTime'][i].split(':')))])
    df_final['PSQI_BedTime'][i] = x

<ipython-input-51-a73b152aec64>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['PSQI_BedTime'][i] = x
<ipython-input-51-a73b152aec64>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['PSQI_BedTime'][i] = x
<ipython-input-51-a73b152aec64>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['PSQI_BedTime'][i] = x
<ipython-input-51-a73b152aec64>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of

In [ ]:
df_final

Subject   Age  PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  PMAT24_A_RTCR  \
0      100206  28.0       81000.0         20.0          0.0        18574.5   
1      100307  28.0       79200.0         17.0          2.0        11839.0   
2      100408  33.0       79200.0          7.0         12.0         3042.0   
3      100610  28.0       77400.0         23.0          0.0        12280.0   
4      101006  33.0       82800.0         11.0          8.0         6569.0   
...       ...   ...           ...          ...          ...            ...   
1081   992774  33.0       79200.0         14.0          5.0        15334.5   
1082   993675  28.0       79200.0         21.0          0.0        22151.0   
1083   994273  28.0        7200.0         20.0          0.0        33421.0   
1084   995174  23.5       82800.0         19.0          0.0        18161.0   
1085   996782  28.0       84600.0         20.0          0.0        18360.0   

      DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  DDisc_SV_3yr_200  \
0                78.13             34.38              9.38             21.88   
1               103.13             46.88            103.13             21.88   
2               153.13             46.88             46.88             46.88   
3               196.88            196.88            184.38            171.88   
4               140.63             96.88            115.63             46.88   
...                ...               ...               ...               ...   
1081              3.13              9.38              3.13              3.13   
1082            159.38            165.63            146.88             96.88   
1083            153.13             21.88              3.13             21.88   
1084            190.63            103.13            121.88             90.63   
1085            196.88            159.38            159.38             96.88   

      ...  Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0     ...    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1     ...    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2     ...    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3     ...    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4     ...    5856.38     5877.30  13309.15216   14458.50988   6514.710165   
...   ...        ...         ...          ...           ...           ...   
1081  ...    6253.92     6234.37  13763.48378   14650.11549   6814.085943   
1082  ...    5994.61     5971.29  12477.92590   13833.70956   6601.958682   
1083  ...    6711.82     6886.75  14637.99165   16242.63411   7100.454111   
1084  ...    6766.36     6734.46  14516.88893   15906.36860   7263.501869   
1085  ...    7095.64     7137.14  14794.60156   15931.95362   7368.394110   

      thal_right_SA   gp_left_SA  gp_right_SA      F      M  
0       8394.220729  3603.322526  3751.440646  False   True  
1       6969.617635  3149.392096  3156.322277   True  False  
2       7763.609633  3476.901847  3725.802635  False   True  
3       8316.436428  3862.828117  3893.266274  False   True  
4       7003.367310  3586.431707  3773.869259   True  False  
...             ...          ...          ...    ...    ...  
1081    7170.076369  3198.083973  3262.912737  False   True  
1082    7131.259938  2995.664299  3186.251544   True  False  
1083    7637.944904  3441.358390  3549.415428  False   True  
1084    7809.016115  3535.013585  3761.052946  False   True  
1085    7834.162855  3319.287200  3451.372013   True  False  

[1086 rows x 35 columns]

#### 4.2.3 Other strings that often crop up in dataframes and need to be replaced with numbers!
`
df_final = df_final.replace('FALSE',0)
df_final = df_final.replace('TRUE',1)
df_final = df_final.replace(False,0)
df_final = df_final.replace(True,1)
df_final = df_final.replace('0',0) # example of random spaces
df_final = df_final.replace(' ',np.NaN) # example of random spaces
`

#### 4.2.4 Making sure every column is of type float
The next line of code will throw an error if you forgot to replace any strings, and it will tell you what those strings are.

In [ ]:
#astype(float) convers all number in the col to float
for col in df_final.columns:
    df_final[col] = df_final[col].astype(float)

In [ ]:
df_final.head()

Subject   Age  PSQI_BedTime  PMAT24_A_CR  PMAT24_A_SI  PMAT24_A_RTCR  \
0  100206.0  28.0       81000.0         20.0          0.0        18574.5   
1  100307.0  28.0       79200.0         17.0          2.0        11839.0   
2  100408.0  33.0       79200.0          7.0         12.0         3042.0   
3  100610.0  28.0       77400.0         23.0          0.0        12280.0   
4  101006.0  33.0       82800.0         11.0          8.0         6569.0   

   DDisc_SV_1mo_200  DDisc_SV_6mo_200  DDisc_SV_1yr_200  DDisc_SV_3yr_200  \
0             78.13             34.38              9.38             21.88   
1            103.13             46.88            103.13             21.88   
2            153.13             46.88             46.88             46.88   
3            196.88            196.88            184.38            171.88   
4            140.63             96.88            115.63             46.88   

   ...  Thal_Left  Thal_Right  str_left_SA  str_right_SA  thal_left_SA  \
0  ...    7827.26     7750.77  15156.32547   16506.40087   7902.211643   
1  ...    5743.19     5773.38  12555.09774   13845.51983   6368.274739   
2  ...    6865.83     6842.16  14165.48884   15590.63180   7235.193241   
3  ...    7452.36     7598.48  15493.70123   16940.05756   7594.851693   
4  ...    5856.38     5877.30  13309.15216   14458.50988   6514.710165   

   thal_right_SA   gp_left_SA  gp_right_SA    F    M  
0    8394.220729  3603.322526  3751.440646  0.0  1.0  
1    6969.617635  3149.392096  3156.322277  1.0  0.0  
2    7763.609633  3476.901847  3725.802635  0.0  1.0  
3    8316.436428  3862.828117  3893.266274  0.0  1.0  
4    7003.367310  3586.431707  3773.869259  1.0  0.0  

[5 rows x 35 columns]

## 5. Handling not available (NA) and inf data:

Sometimes, Python will convert some of your values to + or - infinity, which will result in downstream errors. Convert them to NA, and then handle them as NA values.

In [ ]:
df_final = df_final.replace([np.inf, -np.inf], np.nan)

Next, you need to deal with your NA values. How many nas do you have?

In [ ]:
df_final['DDisc_SV_1mo_200'].isna().sum()

np.int64(8)

In [ ]:
df_final.isna().sum()

Subject              0
Age                  0
PSQI_BedTime         0
PMAT24_A_CR          9
PMAT24_A_SI          9
PMAT24_A_RTCR        9
DDisc_SV_1mo_200     8
DDisc_SV_6mo_200     8
DDisc_SV_1yr_200     8
DDisc_SV_3yr_200     8
DDisc_SV_5yr_200     8
DDisc_SV_10yr_200    8
DDisc_SV_1mo_40K     8
DDisc_SV_6mo_40K     8
DDisc_SV_1yr_40K     8
DDisc_SV_3yr_40K     8
DDisc_SV_5yr_40K     8
DDisc_SV_10yr_40K    8
DDisc_AUC_200        8
DDisc_AUC_40K        8
TBV                  0
Gp_Left              0
Gp_Right             0
Str_Left             0
Str_Right            0
Thal_Left            0
Thal_Right           0
str_left_SA          0
str_right_SA         0
thal_left_SA         0
thal_right_SA        0
gp_left_SA           0
gp_right_SA          0
F                    0
M                    0
dtype: int64

There is a variety of ways to handle `na` data. The most simple approach is to replace `na` data with the median (or mean) value of the feature of interest. There are [other](https://towardsdatascience.com/6-different-ways-to-compensate-for-missing-values-data-imputation-with-examples-6022d9ca0779) [more](https://arxiv.org/abs/1804.11087) sophisticated data imputation techniques out there, many of which actually leverage machine learning tools (so meta)!

However, if a feature has too many NAs, you may want to remove it completely. Define a threshold for the minimal number of missing values that qualifies a feature for removal from your dataset. Here, 12 is quite stringent.

In [ ]:
threshold=12

remove_cols = []
for i in range(len(df_final.columns)):
    if (df_final.iloc[:,i].isnull().sum() >= threshold):
        remove_cols.append(df_final.columns[i])
df_final = df_final.drop(columns=remove_cols)

Next, let's replace the NA values we have left with the feature-specific median (the median is more robust against outliers than the mean is).

In [ ]:
for col in df_final.columns:
    df_final[col].fillna(df_final[col].median(), inplace=True)

In [ ]:
df_final.isna().sum().sum()
# note the difference between df_final.isna().sum() and df_final.isna().sum().sum()

## 6. Removing columns with a standard deviation of 0:

In [ ]:
df_final.std()

In [ ]:
df_final = df_final.loc[:, df_final.std() > 0]

## 7. Feature scaling

You usually need to perform some sort of feature scaling to make sure that all of your variables are in the same range (this affects gradient-descent-based algorithms and distance-based algorithms==).

**Min-Max Scaling / Normalization:** X' = (X-Xmin) / (Xmax-Xmin), X' always ends up with a range of \[0,1\] \
**Standardization / Standard Scaler / Z-score):** X' = (X-mu)/sigma

Which to use? Depends on your data! ["Normalization is good to use when you know that the distribution of your data does not follow a Gaussian distribution. Standardization, on the other hand, can be helpful in cases where the data follows a Gaussian distribution."](https://www.analyticsvidhya.com/blog/2020/04/feature-scaling-machine-learning-normalization-standardization/). Other [popular scaling techniques](https://www.analyticsvidhya.com/blog/2020/07/types-of-feature-transformation-and-scaling/) include the log transform (you often see this with GWAS, ie genome-wide association studies) and dividing your column-wise values by the absolute value of the maximal value of each column (max abs scaler).

Example 1: Min-Max Scaling

In [ ]:
minMaxScaled_df_final=(df_final-df_final.min())/(df_final.max()-df_final.min())

Example 2: Standard Scaling

In [ ]:
standardized_df_final=(df_final-df_final.mean())/df_final.std()

Example 3: Sklearn Min-Max Scaler\
Slightly different from the Min-Max Scaling defined above:\
`
X_std = (X - X.min(axis=0)) / (X.max(axis=0) - X.min(axis=0))
X_scaled = X_std * (max - min) + min
`

In [ ]:
from sklearn.preprocessing import MinMaxScaler
mms=MinMaxScaler()
mms.fit(df_final)
df_final_mms=mms.transform(df_final)

# 8. Data visualization

Data visualization is a wonderful way to get to know your data in order to plan a relevant analysis or find an appropriate machine learning application. [Matplotlib](https://matplotlib.org/) and [seaborn](https://seaborn.pydata.org/) are two canonical data visualizations tools that you can use in Python.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

#### 8.1 Scatter plot

In [ ]:
plt.scatter(df_final["PSQI_BedTime"],df_final["TBV"])
plt.ylabel("Total brain volume (mm3)")
plt.xlabel("Bedtime (seconds)")
plt.title("Total brain volume as a function of bed time")

#### 8.2 [Violin plot](https://chartio.com/learn/charts/violin-plot-complete-guide/)

In [ ]:
plt.violinplot(df_final['PSQI_BedTime'])
plt.ylabel("Bed time")
plt.title("Violin plot of bed times in our sample")

#### 8.3 Histogram plot

In [ ]:
sns.displot(df_final,x='TBV',kind='kde',fill=True) # smoothed histogram
plt.ylabel("Density")
plt.xlabel("Total brain volume (mm3)")
plt.title("Distribution of total brain volume in our sample")

In [ ]:
sns.histplot(df_final,x='Str_Left',fill=True,color='orange')
sns.histplot(df_final,x='Thal_Left',fill=True,color='turquoise')
plt.legend(['Left striatum','Left thalamus'])
plt.ylabel("Density")
plt.xlabel("Structure-specific volume (mm3)")
plt.title("Distribution of different structure volumes in our sample")
plt.show()

# 9. Loading local files into Google Colab

Nice tutorial [here](https://neptune.ai/blog/google-colab-dealing-with-files)

**TLDR:**

In [ ]:
from google.colab import files # files module from google.colab library

In [ ]:
uploaded = files.upload # you can then select files and upload them to your colab

# 10. Mini data visualization assignment
PS: this [list of named colours](https://matplotlib.org/3.5.0/gallery/color/named_colors.html) can help make your graphs nicer :)
<ol>
<li>Generate a scatter plot of TBV as a function of bed time, by gender.</li>
<li>Generate a violin plot of bed times in our sample for subjects over 30 years.</li>
<li>Generate two smoothed and superimposed histograms of bed times in subjects that are above and below 30 years.</li>
<li>Generate two superimposed histograms of the volume distributions in the left hemisphere (left striatum, thalamus and globus pallidus) and right hemisphere (right striatum, thalamus and globus pallidus)</li>
</ol>

# Supplementary

### Interactive plots with Plotly Python

* FYI: Plotly is based in Montreal!
* [Superb documentation](https://plotly.com/python/)
* Plotly express is more "higher-level" (tons of parameters are already set for you); can save you time but is less flexible

### Bringing it to the next level: making an App with Dash
* Documentation [here](https://dash.plotly.com/)
* html + Python